File: SimpleKeychain/A0SimpleKeychain.m
#import "A0SimpleKeychain.h"

@interface A0SimpleKeychain ()

@end

@implementation A0SimpleKeychain

- (instancetype)init {
    NSString *service = [[NSBundle mainBundle] bundleIdentifier];
    return [self initWithService:service accessGroup:nil];
}

- (instancetype)initWithService:(NSString *)service {
    return [self initWithService:service accessGroup:nil];
}

- (instancetype)initWithService:(NSString *)service accessGroup:(NSString *)accessGroup {
    self = [super init];
    if (self) {
        _service = service;
        _accessGroup = accessGroup;
        _defaultAccessiblity = A0SimpleKeychainItemAccessibleAfterFirstUnlock;
        _useAccessControl = NO;
        
// This does not apply to watchOS & tvOS
#if A0LocalAuthenticationCapable
        _localAuthenticationContext = [LAContext new];
        _localAuthenticationContext.touchIDAuthenticationAllowableReuseDuration = 0;
#endif
    }
    return self;
}

- (void)setTouchIDAuthenticationAllowableReuseDuration:(NSTimeInterval) duration {
// This does not apply to watchOS & tvOS
#if A0LocalAuthenticationCapable
    if (duration <= 0) {
        _localAuthenticationContext.touchIDAuthenticationAllowableReuseDuration = 0;
    } else if (duration >= LATouchIDAuthenticationMaximumAllowableReuseDuration) {
        _localAuthenticationContext.touchIDAuthenticationAllowableReuseDuration = LATouchIDAuthenticationMaximumAllowableReuseDuration;
    } else {
        _localAuthenticationContext.touchIDAuthenticationAllowableReuseDuration = duration;
    }
#endif
}

- (NSString *)stringForKey:(NSString *)key {
    return [self stringForKey:key promptMessage:nil];
}

- (NSData *)dataForKey:(NSString *)key {
    return [self dataForKey:key promptMessage:nil];
}

- (NSString *)stringForKey:(NSString *)key promptMessage:(NSString *)message {
    NSData *data = [self dataForKey:key promptMessage:message];
    NSString *string = nil;
    if (data) {
        string = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    }
    return string;
}

- (NSData *)dataForKey:(NSString *)key promptMessage:(NSString *)message {
    return [self dataForKey:key promptMessage:message error:nil];
}

- (NSData *)dataForKey:(NSString *)key promptMessage:(NSString *)message error:(NSError**)err {
    if (!key) {
        return nil;
    }
    
    NSDictionary *query = [self queryFetchOneByKey:key message:message];
    CFTypeRef data = nil;
    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, &data);
    if (status != errSecSuccess) {
        if(err != nil) {
            *err = [NSError errorWithDomain:A0ErrorDomain code:status userInfo:@{NSLocalizedDescriptionKey : [self stringForSecStatus:status]}];
        }
        return nil;
    }
    
    NSData *dataFound = [NSData dataWithData:(__bridge NSData *)data];
    if (data) {
        CFRelease(data);
    }
    
    return dataFound;
}

- (BOOL)hasValueForKey:(NSString *)key {
    if (!key) {
        return NO;
    }
    NSDictionary *query = [self queryFindByKey:key message:nil];
    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, NULL);
    return status == errSecSuccess;
}

- (nonnull NSArray *)keys {
    NSMutableArray *keys = [NSMutableArray array];
    NSDictionary *query = [self queryFindAll];
    CFArrayRef result = nil;
    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, (CFTypeRef *)&result);
    if (status == errSecSuccess) {
        NSArray *items = [NSArray arrayWithArray:(__bridge NSArray *)result];
        CFBridgingRelease(result);
        for (NSDictionary *item in items) {
            id secAccount = item[(__bridge id)kSecAttrAccount];
            if ([secAccount isKindOfClass:[NSString class]]) {
                NSString *key = (NSString *)secAccount;
                [keys addObject:key];
            }
        }
    }
    return keys;
}

- (BOOL)setString:(NSString *)string forKey:(NSString *)key {
    return [self setString:string forKey:key promptMessage:nil];
}

- (BOOL)setData:(NSData *)data forKey:(NSString *)key {
    return [self setData:data forKey:key promptMessage:nil];
}

- (BOOL)setString:(NSString *)string forKey:(NSString *)key promptMessage:(NSString *)message {
    NSData *data = key ? [string dataUsingEncoding:NSUTF8StringEncoding] : nil;
    return [self setData:data forKey:key promptMessage:message];
}


- (BOOL)setData:(NSData *)data forKey:(NSString *)key promptMessage:(NSString *)message {
    if (!key) {
        return NO;
    }
    
    NSDictionary *query = [self queryFindByKey:key message:message];

    // Touch ID case
    if (self.useAccessControl && self.defaultAccessiblity == A0SimpleKeychainItemAccessibleWhenPasscodeSetThisDeviceOnly) {
        // TouchId case. Doesn't support updating keychain items
        // see Known Issues: https://developer.apple.com/library/ios/releasenotes/General/RN-iOSSDK-8.0/
        // We need to delete old and add a new item. This can fail
        OSStatus status = SecItemDelete((__bridge CFDictionaryRef)query);
        if (status == errSecSuccess || status == errSecItemNotFound) {
            NSDictionary *newQuery = [self queryNewKey:key value:data];
            OSStatus status = SecItemAdd((__bridge CFDictionaryRef)newQuery, NULL);
            return status == errSecSuccess;
        }
    }

    // Normal case
    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, NULL);
    if (status == errSecSuccess) {
        if (data) {
            NSDictionary *updateQuery = [self queryUpdateValue:data message:message];
            status = SecItemUpdate((__bridge CFDictionaryRef)query, (__bridge CFDictionaryRef)updateQuery);
            return status == errSecSuccess;
        } else {
            OSStatus status = SecItemDelete((__bridge CFDictionaryRef)query);
            return status == errSecSuccess;
        }
    } else {
        NSDictionary *newQuery = [self queryNewKey:key value:data];
        OSStatus status = SecItemAdd((__bridge CFDictionaryRef)newQuery, NULL);
        return status == errSecSuccess;
    }
}

- (BOOL)deleteEntryForKey:(NSString *)key {
    if (!key) {
        return NO;
    }
    NSDictionary *deleteQuery = [self queryFindByKey:key message:nil];
    OSStatus status = SecItemDelete((__bridge CFDictionaryRef)deleteQuery);
    return status == errSecSuccess;
}

- (void)clearAll {
  NSMutableDictionary *queryDelete = [self baseQuery];
  queryDelete[(__bridge id)kSecClass] = (__bridge id)kSecClassGenericPassword;
#if !TARGET_OS_IPHONE
  queryDelete[(__bridge id)kSecMatchLimit] = (__bridge id)kSecMatchLimitAll;
#endif
  OSStatus status = SecItemDelete((__bridge CFDictionaryRef)queryDelete);
  if (status != errSecSuccess) {
    return;
  }
}

+ (A0SimpleKeychain *)keychain {
    return [[A0SimpleKeychain alloc] init];
}

+ (A0SimpleKeychain *)keychainWithService:(NSString *)service {
    return [[A0SimpleKeychain alloc] initWithService:service];
}

+ (A0SimpleKeychain *)keychainWithService:(NSString *)service accessGroup:(NSString *)accessGroup {
    return [[A0SimpleKeychain alloc] initWithService:service accessGroup:accessGroup];
}

#pragma mark - Utility methods

- (CFTypeRef)accessibility {
    CFTypeRef accessibility;
    switch (self.defaultAccessiblity) {
        case A0SimpleKeychainItemAccessibleAfterFirstUnlock:
            accessibility = kSecAttrAccessibleAfterFirstUnlock;
            break;
        case A0SimpleKeychainItemAccessibleAfterFirstUnlockThisDeviceOnly:
            accessibility = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly;
            break;
        case A0SimpleKeychainItemAccessibleWhenPasscodeSetThisDeviceOnly:
            accessibility = kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly;
            break;
        case A0SimpleKeychainItemAccessibleWhenUnlocked:
            accessibility = kSecAttrAccessibleWhenUnlocked;
            break;
        case A0SimpleKeychainItemAccessibleWhenUnlockedThisDeviceOnly:
            accessibility = kSecAttrAccessibleWhenUnlockedThisDeviceOnly;
            break;
        default:
            accessibility = kSecAttrAccessibleWhenUnlockedThisDeviceOnly;
    }
    return accessibility;
}

- (NSString*)stringForSecStatus:(OSStatus)status {
    
    switch(status) {
        case errSecSuccess:
            return NSLocalizedStringFromTable(@"errSecSuccess: No error", @"SimpleKeychain", @"Possible error from keychain. ");
        case errSecUnimplemented:
            return NSLocalizedStringFromTable(@"errSecUnimplemented: Function or operation not implemented", @"SimpleKeychain", @"Possible error from keychain. ");
        case errSecParam:
            return NSLocalizedStringFromTable(@"errSecParam: One or more parameters passed to the function were not valid", @"SimpleKeychain", @"Possible error from keychain. ");
        case errSecAllocate:
            return NSLocalizedStringFromTable(@"errSecAllocate: Failed to allocate memory", @"SimpleKeychain", @"Possible error from keychain. ");
        case errSecNotAvailable:
            return NSLocalizedStringFromTable(@"errSecNotAvailable: No trust results are available", @"SimpleKeychain", @"Possible error from keychain. ");
        case errSecAuthFailed:
            return NSLocalizedStringFromTable(@"errSecAuthFailed: Authorization/Authentication failed", @"SimpleKeychain", @"Possible error from keychain. ");
        case errSecDuplicateItem:
            return NSLocalizedStringFromTable(@"errSecDuplicateItem: The item already exists", @"SimpleKeychain", @"Possible error from keychain. ");
        case errSecItemNotFound:
            return NSLocalizedStringFromTable(@"errSecItemNotFound: The item cannot be found", @"SimpleKeychain", @"Possible error from keychain. ");
        case errSecInteractionNotAllowed:
            return NSLocalizedStringFromTable(@"errSecInteractionNotAllowed: Interaction with the Security Server is not allowed", @"SimpleKeychain", @"Possible error from keychain. ");
        case errSecDecode:
            return NSLocalizedStringFromTable(@"errSecDecode: Unable to decode the provided data", @"SimpleKeychain", @"Possible error from keychain. ");
        default:
            return [NSString stringWithFormat:NSLocalizedStringFromTable(@"Unknown error code %d", @"SimpleKeychain", @"Possible error from keychain. "), (int)status];
    }
}

#pragma mark - Query Dictionary Builder methods

- (NSMutableDictionary *)baseQuery {
    NSMutableDictionary *attributes = [@{
                                         (__bridge id)kSecClass: (__bridge id)kSecClassGenericPassword,
                                         (__bridge id)kSecAttrService: self.service,
                                         } mutableCopy];

#if !TARGET_IPHONE_SIMULATOR
    if (self.accessGroup) {
        attributes[(__bridge id)kSecAttrAccessGroup] = self.accessGroup;
    }
    
#if A0LocalAuthenticationCapable
    attributes[(__bridge id)kSecUseAuthenticationContext] = self.localAuthenticationContext;
#endif
#endif

    return attributes;
}

- (NSDictionary *)queryFindAll {
    NSMutableDictionary *query = [self baseQuery];
    [query addEntriesFromDictionary:@{
                                     (__bridge id)kSecReturnAttributes: @YES,
                                     (__bridge id)kSecMatchLimit: (__bridge id)kSecMatchLimitAll,
                                     }];
    return query;
}

- (NSDictionary *)queryFindByKey:(NSString *)key message:(NSString *)message {
    NSAssert(key != nil, @"Must have a valid non-nil key");
    NSMutableDictionary *query = [self baseQuery];
    query[(__bridge id)kSecAttrAccount] = key;
    if (message) {
        query[(__bridge id)kSecUseOperationPrompt] = message;
    }
    return query;
}

- (NSDictionary *)queryUpdateValue:(NSData *)data message:(NSString *)message {
    if (message) {
        return @{
            (__bridge id)kSecUseOperationPrompt: message,
            (__bridge id)kSecValueData: data
        };
    }
    return @{(__bridge id)kSecValueData: data};
}

- (NSDictionary *)queryNewKey:(NSString *)key value:(NSData *)value {
    NSMutableDictionary *query = [self baseQuery];
    query[(__bridge id)kSecAttrAccount] = key;
    query[(__bridge id)kSecValueData] = value;
    if (self.useAccessControl) {
        CFErrorRef error = NULL;
        SecAccessControlRef accessControl = SecAccessControlCreateWithFlags(kCFAllocatorDefault, [self accessibility], kSecAccessControlUserPresence, &error);
        if (error == NULL || accessControl != NULL) {
            query[(__bridge id)kSecAttrAccessControl] = (__bridge_transfer id)accessControl;
            query[(__bridge id)kSecUseAuthenticationUI] = (__bridge_transfer id)kSecUseAuthenticationUIFail;
        }
    } else {
        query[(__bridge id)kSecAttrAccessible] = (__bridge id)[self accessibility];
    }
    return query;
}

- (NSDictionary *)queryFetchOneByKey:(NSString *)key message:(NSString *)message {
    NSMutableDictionary *query = [self baseQuery];
    [query addEntriesFromDictionary:@{
                                      (__bridge id)kSecReturnData: @YES,
                                      (__bridge id)kSecMatchLimit: (__bridge id)kSecMatchLimitOne,
                                      (__bridge id)kSecAttrAccount: key,
                                      }];
    if (self.useAccessControl && message) {
        query[(__bridge id)kSecUseOperationPrompt] = message;
    }

    return query;
}
@end

File: SimpleKeychain/SimpleKeychain/A0SimpleKeychain.swift
// A0SimpleKeychain.swift
//
// Copyright (c) 2015 Auth0 (http://auth0.com)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

import Foundation

/**
*  Enum with Kechain items accessibility types. It's a mirror of `kSecAttrAccessible` values.
*/
enum A0SimpleKeychainItemAccessible {
    /**
    *  @see kSecAttrAccessibleWhenUnlocked
    */
    case WhenUnlocked
    /**
    *  @see kSecAttrAccessibleAfterFirstUnlock
    */
    case AfterFirstUnlock
    /**
    *  @see kSecAttrAccessibleAlways
    */
    case Always
    /**
    *  @see kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly
    */
    case WhenPasscodeSetThisDeviceOnly
    /**
    *  @see kSecAttrAccessibleWhenUnlockedThisDeviceOnly
    */
    case WhenUnlockedThisDeviceOnly
    /**
    *  kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
    */
    case AfterFirstUnlockThisDeviceOnly
    /**
    *  @see kSecAttrAccessibleAlwaysThisDeviceOnly
    */
    case AlwaysThisDeviceOnly
}

/**
* Enum with keychain error codes. It's a mirror of the keychain error codes.
*/
enum A0SimpleKeychainErrorCode: Int {
    /**
    * @see errSecSuccess
    */
    case NoError = 0
    /**
    * @see errSecUnimplemented
    */
    case Unimplemented = -4
    /**
    * @see errSecParam
    */
    case WrongParameter = -50
    /**
    * @see errSecAllocate
    */
    case Allocation = -108
    /**
    * @see errSecNotAvailable
    */
    case NotAvailable = -25291
    /**
    * @see errSecAuthFailed
    */
    case AuthFailed = -25293
    /**
    * @see errSecDuplicateItem
    */
    case DuplicateItem = -25299
    /**
    * @see errSecItemNotFound
    */
    case ItemNotFound = -25300
    /**
    * @see errSecInteractionNotAllowed
    */
    case InteractionNotAllowed = -25308
    /**
    * @see errSecDecode
    */
    case Decode = -26275
}

/**
*  A simple helper class to deal with storing and retrieving values from iOS Keychain.
*  It has support for sharing keychain items using Access Group and also for iOS 8 fine grained accesibility over a specific Kyechain Item (Using Access Control).
*  The support is only available for iOS 8+, otherwise it will default using the coarse grained accesibility field.
*  When a `NSString` or `NSData` is stored using Access Control and the accesibility flag `A0SimpleKeychainItemAccessible.WhenPasscodeSetThisDeviceOnly`, iOS will prompt the user for it's passcode or pass a TouchID challenge (if available).
*/
class A0SimpleKeychain: NSObject {

    private let errorDomain = "com.auth0.simplekeychain"
    private let localizedTableName = "SimpleKeychain"

    /**
    *  Service name under all items are saved. Default value is Bundle Identifier.
    */
    private(set) var service: String

    /**
    *  Access Group for Keychain item sharing. If it's nil no keychain sharing is possible. Default value is nil.
    */
    private(set) var accessGroup: String?

    /**
    *  What type of accessibility the items stored will have. All values are translated to `kSecAttrAccessible` constants.
    *  Default value is A0SimpleKeychainItemAccessibleAfterFirstUnlock.
    *  @see kSecAttrAccessible
    */
    var defaultAccessibility: A0SimpleKeychainItemAccessible
    /**
    *  Tells A0SimpleKeychain to use `kSecAttrAccessControl` instead of `kSecAttrAccessible`. It will work only in iOS 8+, defaulting to `kSecAttrAccessible` on lower version.
    *  Default value is NO.
    */
    var useAccessControl: Bool

    ///---------------------------------------------------
    /// @name Initialization
    ///---------------------------------------------------

    /**
    Creates an instance of A0SimpleKeychain

    :param: service              name of the service.
    :param: accessGroup          name of the access group used for keychain sharing. By default is nil.
    :param: defaultAccessibility default accessibility value for keychain entries. By default is .AfterFirstUnlock.
    :param: useAccessControl     if iOS 8 Access Control feature is used. By default is false.

    :returns: a new instance
    */
    init(service: String? = NSBundle.mainBundle().bundleIdentifier, accessGroup: String? = nil, defaultAccessibility: A0SimpleKeychainItemAccessible = .AfterFirstUnlock, useAccessControl: Bool = false) {
        self.service = service!
        self.accessGroup = accessGroup
        self.defaultAccessibility = defaultAccessibility
        self.useAccessControl = useAccessControl
    }

    ///---------------------------------------------------
    /// @name Store values
    ///---------------------------------------------------

    /**
    Stores a string in the keychain

    :param: string  string value to store
    :param: key     key for the string value
    :param: message prompt message displayed if TouchID integration is enabled. By default is nil.

    :returns: if the value was stored successfully
    */
    func setString(string: String, forKey key: String, promptMessage message: String? = nil) -> Bool {
        if let data = string.dataUsingEncoding(NSUTF8StringEncoding) {
            return setData(data, forKey: key, promptMessage: message)
        }
        return false
    }

    /**
    Stores a NSData in the keychain

    :param: data    data to store
    :param: key     key for the data value
    :param: message prompt message displayed if TouchID integration is enabled. By default is nil.

    :returns: if the value was stored successfully
    */
    func setData(data: NSData, forKey key: String, promptMessage message: String? = nil) -> Bool {
        let query = queryFindByKey(key, message: message)
        // Touch ID case
        if (self.useAccessControl && self.defaultAccessibility == .WhenPasscodeSetThisDeviceOnly) {
            // TouchId case. Doesn't support updating keychain items
            // see Known Issues: https://developer.apple.com/library/ios/releasenotes/General/RN-iOSSDK-8.0/
            // We need to delete old and add a new item. This can fail
            let status = SecItemDelete(query)
            if (status == errSecSuccess || status == errSecItemNotFound) {
                return insertItemWithKey(key, value: data)
            }
        }
        let findStatus = SecItemCopyMatching(query, nil)
        if (findStatus == errSecSuccess) {
            return SecItemUpdate(query, queryUpdateValue(data, message: message)) == errSecSuccess
        } else {
            return SecItemAdd(queryNewItemWithKey(key, value: data), nil) == errSecSuccess
        }
    }

    ///---------------------------------------------------
    /// @name Obtain values
    ///---------------------------------------------------

    /**
    Fetch a string value from the keychain. If you are using swift please use stringForKey(key:, promptMessage:) instead.

    :param: key     key of the value to fetch
    :param: message prompt message displayed if TouchID integration is enabled. By default is nil.
    :param: error   when an error occurs, an error will be stored here otherwise it will be nil.

    :returns: stored string value or nil
    */
    func stringForKey(key: String, promptMessage message: String? = nil, error: NSErrorPointer) -> String? {
        let result = stringForKey(key, promptMessage: message)
        switch(result) {
        case (let string, nil) where string != nil:
            return string
        case (_, let err):
            if(error != nil) {
                error.memory = err
            }
            return nil
        }
    }

    /**
    Fetch a string value from the keychain

    :param: key     key of the value to fetch
    :param: message prompt message displayed if TouchID integration is enabled. By default is nil.

    :returns: a tuple with the string value if successful otherwise with the error
    */
    func stringForKey(key: String, promptMessage message: String? = nil) -> (string: String?, error: NSError?) {
        let result = dataForKey(key, promptMessage: message)
        switch(result) {
        case (let data, nil) where data != nil:
            return (string: NSString(data: data!, encoding: NSUTF8StringEncoding), error: nil)
        case (_, let error):
            return (string: nil, error: error)
        }
    }

    /**
    Fetch a NSData value from the keychain. If you are using swift please use dataForKey(key:, promptMessage:) instead.

    :param: key     key of the value to fetch
    :param: message prompt message displayed if TouchID integration is enabled. By default is nil.
    :param: error   when an error occurs, an error will be stored here otherwise it will be nil.

    :returns: stored NSData value or nil
    */
    func dataForKey(key: String, promptMessage message: String? = nil, error: NSErrorPointer) -> NSData? {
        let result = dataForKey(key, promptMessage: message)
        switch (result) {
        case (let data, nil) where data != nil:
            return data;
        case (_, let err):
            if (error != nil) {
                error.memory = err
            }
            return nil
        }
    }

    /**
    Fetch a NSData value from the keychain

    :param: key     key of the value to fetch
    :param: message prompt message displayed if TouchID integration is enabled. By default is nil.

    :returns: a tuple with the NSData value if successful otherwise with the error
    */
    func dataForKey(key: String, promptMessage message: String? = nil) -> (data: NSData?, error: NSError?) {
        let query = querySingleItemByKey(key, message: message)
        var data : Unmanaged<AnyObject>?
        let status = SecItemCopyMatching(query, &data)
        if (status != errSecSuccess) {
            let code = Int(status)
            return (data: nil, error: NSError(domain: errorDomain, code: code, userInfo: [NSLocalizedDescriptionKey: localizedErrorFromStatus(status)]))
        }

        return (data: data?.takeUnretainedValue() as? NSData, nil)
    }

    /**
    Check if there is an entry in the keychain with the given key

    :param: key key to find

    :returns: if the entry exists or not
    */
    func hasValueForKey(key: String) -> Bool {
        let query = queryFindByKey(key, message: nil)
        return SecItemCopyMatching(query, nil) == errSecSuccess
    }

    ///---------------------------------------------------
    /// @name Remove values
    ///---------------------------------------------------

    /**
    Removes an entry from the keychain

    :param: key     entry key to remove
    :param: message message displayed to confirm if required

    :returns: if the entry was removed
    */
    func removeValueForKey(key: String, promptMessage message: String? = nil) -> Bool {
        let query = queryFindByKey(key, message: message)
        return SecItemDelete(query) == errSecSuccess
    }

    /**
    Remove all entries from the kechain with the service and access group values.
    */
    func removeAllValues() {
        let query = queryFindAll()
        var result: Unmanaged<AnyObject>?
        let status = SecItemCopyMatching(query, &result)
        if (status == errSecSuccess || status == errSecItemNotFound) {
            if let items = result?.takeUnretainedValue() as? NSArray {
                for item in items {
                    if let dict = item as? NSDictionary {
                        var query = NSDictionary(dictionary: dict)
                        query.setValue(NSString(format:kSecClassGenericPassword), forKey: NSString(format:kSecClass))
                        SecItemDelete(query)
                    }
                }
            }
        }
    }

    // MARK: - Query Builders

    private func insertItemWithKey(key: String, value: NSData) -> Bool {
        let queryNewItem = queryNewItemWithKey(key, value: value)
        let status = SecItemAdd(queryNewItem, nil)
        return status == errSecSuccess
    }

    private func baseQuery() -> [NSString: AnyObject] {
        var attributes = [
            NSString(format: kSecClass): NSString(format: kSecClassGenericPassword),
            NSString(format: kSecAttrService): self.service
        ]
#if !TARGET_IPHONE_SIMULATOR
        if let accessGroupName = self.accessGroup {
            attributes[NSString(format: kSecAttrAccessGroup)] = accessGroupName
        }
#endif
        return attributes
    }

    private func queryFindByKey(key: String, message: String?)  -> [NSString: AnyObject] {
        var query = baseQuery()
        query[NSString(format: kSecAttrAccount)] = key
        if let operationPrompt = message {
            query[NSString(format: kSecUseOperationPrompt)] = operationPrompt
        }
        return query
    }

    private func queryNewItemWithKey(key: String, value: NSData) -> [NSString: AnyObject] {
        var query = baseQuery()
        query[NSString(format:kSecAttrAccount)] = key
        query[NSString(format:kSecValueData)] = value
        if (self.useAccessControl && floor(NSFoundationVersionNumber) > NSFoundationVersionNumber_iOS_7_1) {
            var error: Unmanaged<CFError>?
            var accessControl = SecAccessControlCreateWithFlags(kCFAllocatorDefault, accessibility(), .UserPresence, &error)
            if (error == nil || accessControl != nil) {
                query[NSString(format:kSecAttrAccessControl)] = accessControl.takeUnretainedValue()
                query[NSString(format:kSecUseNoAuthenticationUI)] = true
            }
        } else {
            query[NSString(format:kSecAttrAccessible)] = accessibility()
        }
        return query
    }

    private func queryUpdateValue(value: NSData, message: String?) -> [NSString : AnyObject] {
        var query : [NSString : AnyObject] = [
            NSString(format: kSecValueData): value
        ]
        if let promptMessage = message {
            query[NSString(format:kSecUseOperationPrompt)] = promptMessage
        }
        return query
    }

    private func querySingleItemByKey(key: String, message: String?) -> [NSString : AnyObject] {
        var query = baseQuery()
        query[NSString(format:kSecReturnData)] = true
        query[NSString(format:kSecMatchLimit)] = NSString(format: kSecMatchLimitOne)
        query[NSString(format:kSecAttrAccount)] = key
        if let promptMessage = message {
            query[NSString(format:kSecUseOperationPrompt)] = promptMessage
        }
        return query
    }

    private func queryFindAll() -> [NSString : AnyObject] {
        var query = [
            NSString(format:kSecReturnAttributes) : true,
            NSString(format:kSecMatchLimit) : NSString(format:kSecMatchLimitAll)
        ]
        return query;
    }

    //MARK: - Utility methods

    private func accessibility() -> CFTypeRef {
        switch(self.defaultAccessibility) {
        case .AfterFirstUnlock:
            return kSecAttrAccessibleAfterFirstUnlock
        case .Always:
            return kSecAttrAccessibleAlways
        case .AfterFirstUnlockThisDeviceOnly:
            return kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        case .AlwaysThisDeviceOnly:
            return kSecAttrAccessibleAlwaysThisDeviceOnly
        case .WhenPasscodeSetThisDeviceOnly:
            if (floor(NSFoundationVersionNumber) > NSFoundationVersionNumber_iOS_7_1) { //iOS 8
                return kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly
            } else { //iOS <= 7.1
                return kSecAttrAccessibleWhenUnlockedThisDeviceOnly
            }
        case .WhenUnlocked:
            return kSecAttrAccessibleWhenUnlocked
        case .WhenUnlockedThisDeviceOnly:
            return kSecAttrAccessibleWhenUnlockedThisDeviceOnly
        }
    }

    private func localizedErrorFromStatus(status: OSStatus) -> String {
        switch(status) {
        case errSecSuccess:
            return NSLocalizedString("errSecSuccess: No error", tableName: localizedTableName, comment: "Possible error from keychain.")
        case errSecUnimplemented:
            return NSLocalizedString("errSecUnimplemented: Function or operation not implemented", tableName: localizedTableName, comment: "Possible error from keychain.")
        case errSecParam:
            return NSLocalizedString("errSecParam: One or more parameters passed to the function were not valid", tableName: localizedTableName, comment: "Possible error from keychain.")
        case errSecAllocate:
            return NSLocalizedString("errSecAllocate: Failed to allocate memory", tableName: localizedTableName, comment: "Possible error from keychain.")
        case errSecNotAvailable:
            return NSLocalizedString("errSecNotAvailable: No trust results are available", tableName: localizedTableName, comment: "Possible error from keychain.")
        case errSecAuthFailed:
            return NSLocalizedString("errSecAuthFailed: Authorization/Authentication failed", tableName: localizedTableName, comment: "Possible error from keychain.")
        case errSecDuplicateItem:
            return NSLocalizedString("errSecDuplicateItem: The item already exists", tableName: localizedTableName, comment: "Possible error from keychain.")
        case errSecItemNotFound:
            return NSLocalizedString("errSecItemNotFound: The item cannot be found", tableName: localizedTableName, comment: "Possible error from keychain.")
        case errSecInteractionNotAllowed:
            return NSLocalizedString("errSecInteractionNotAllowed: Interaction with the Security Server is not allowed", tableName: localizedTableName, comment: "Possible error from keychain.")
        case errSecDecode:
            return NSLocalizedString("errSecDecode: Unable to decode the provided data", tableName: localizedTableName, comment: "Possible error from keychain.")
        default:
            return NSString(format:NSLocalizedString("Unknown error code %d", tableName: localizedTableName, comment: "Possible error from keychain."), status)
        }
    }
}
